### PR TITLE
Rewrite crithack

### PR DESCRIPTION
**Pull request by BenCat07**
### This completely changes crithack, And adds alot of things needed to get it to work.

The biggest change  is that experimental crits are now the standard mode, and that they no longer make you glitch out. The crit logic was also refined so it works way more reliably (still not 100% though).
Other changes include: 
- Displaying why you can't crit anymore
- Displaying how many shots are needed to crit again
- Adding a new type of CreateMove made for modifying command_number

TODO:
- [x] Make m_iCrits from DT_TFPlayerScoringDataExclusive get used in order to prevent fake crits from reducing crit bucket
- [x] Rename rvars to to encourage compability with the previous version
- [x] Calculate observed crit chance on client instead of relying on the server
- [x] Possibly make UI more compact
- [x] Fix the First two crits adding/Subtracting from bucket more than once
- [ ] Possibly refactor the code

Images (more soon:tm):
image

### CODE REVIEW

File: src/crits.cpp [hunk 1]
```diff
@@ -1,357 +1,655 @@
...
+            AddCenterString("Forcing Crits!", colors::red);
+
+        // Mark if the weapon can randomly crit or not
+        if (!re::C_TFWeaponBase::CanFireCriticalShot(RAW_ENT(LOCAL_W), false, nullptr))
+        {
+            AddCenterString("Weapon cannot randomly crit.", colors::red);
+            return;
         }
+
+        // Crit chance is not low enough
```
Comment (TotallyNotElite) [inline src/crits.cpp:565]:
```suggestion
        // Observed crit chance is not low enough
```

File: src/crits.cpp [hunk 2]
```diff
@@ -1,357 +1,655 @@
...
+    weapon_info info(wep);
+    // Increment call count
+    int call_count = info.crit_count + 1;
+    // How many times there was a check for crits
+    int crit_checks  = info.crit_attempts + 1;
+    float flMultiply = 0.5;
+
+    if (g_pLocalPlayer->weapon_mode != weapon_melee)
+    {
+        float flTmp = min(max((((float) call_count / (float) crit_checks) - 0.1) * 1.111, 0.0), 1.0);
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:81]: This is code taken from the game directly that i did not want to touch

File: src/crits.cpp [hunk 3]
```diff
@@ -1,357 +1,655 @@
...
+    {
+        // Used later
+        int xor_dat = (weapon->entindex() << 8 | LOCAL_E->m_IDX);
+        // Clear old data
+        crit_cmds.clear();
+        added_per_shot = 0.0f;
 
-        // If backtracking, use the backtrack ticks
-        if (bt::isBacktrackEnabled)
+        // 30.000.000 seems like a good cap
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:338]: Ticks to go into the future at max

File: src/crits.cpp [hunk 4]
```diff
@@ -1,357 +1,655 @@
...
+        for (int i = cur_cmdnum, j = 5; i <= cur_cmdnum + 30000000 + 200 && j > 0; i++)
         {
-            // Loop all ticks
-            for (int j = 0; j < 66; j++)
+            // Reverse seed
+            int iSeed    = MD5_PseudoRandom(i) & 0x7fffffff;
+            int tempSeed = iSeed ^ (xor_dat);
+            RandomSeed(tempSeed);
+            int iResult = RandomInt(0, 9999);
+            // Returns something low? Store it.
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:346]: Why what?

File: src/crits.cpp [hunk 5]
```diff
@@ -1,357 +1,655 @@
...
+void fix_observed_critchance(IClientEntity *weapon)
+{
+    weapon_info info(weapon);
+    // Fix crit chance
+    info.observed_crit_chance = GetObservedCritChane();
+    info.restore_data(weapon);
+}
+
+// Fix bucket on non-local servers
+void unfuck_bucket(IClientEntity *weapon)
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:412]: name taken from old crithack, can change if you want

File: src/crits.cpp [hunk 6]
```diff
@@ -1,357 +1,655 @@
...
+
+    info.crit_bucket = bucket;
+    info.restore_data(weapon);
 }
-void create_move()
+
+void CreateMove()
 {
-    if (!crit_key && !crit_melee && !auto_crit)
+    // Need to do regardless
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:474]: Need to store damage regardless if crithack is on or not

File: src/crits.cpp [hunk 7]
```diff
@@ -1,357 +1,655 @@
...
+        return;
+    logging::Info("Multiplier: %f", GetWithdrawMult(RAW_ENT(LOCAL_W)));
+});
+
+static Timer bucket_cap_refresh{};
+float GetBucketCap()
+{
+    static float return_value;
+    static ConVar *tf_weapon_criticals = g_ICvar->FindVar("tf_weapon_criticals_bucket_cap");
+    if (bucket_cap_refresh.test_and_set(5000))
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:100]: True

File: src/crits.cpp [hunk 8]
```diff
@@ -1,357 +1,655 @@
...
+    return flToRemove;
+}
 
-    crithack_saved_state state{};
-    state.Save(weapon);
+static CatCommand debug_print_mult("debug_print_mult", "print Withdraw multiplier", []() {
+    if (CE_BAD(LOCAL_E) || CE_BAD(LOCAL_W))
+        return;
+    logging::Info("Multiplier: %f", GetWithdrawMult(RAW_ENT(LOCAL_W)));
+});
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:93]: no, thanks for catching that

File: src/crits.cpp [hunk 9]
```diff
@@ -1,357 +1,655 @@
...
+    }
+    float flToRemove = GetWithdrawMult(wep) * flDamage;
+    // Can remove
+    if (flToRemove <= info.crit_bucket)
+        return true;
+
+    return false;
+}
+
+// As opposed to the function above this does actual writing
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:123]: It should not do that, as incrementing crit_count is going to conflict with the game which also increments it

File: src/crits.cpp [hunk 10]
```diff
@@ -1,357 +1,655 @@
...
+        info.crit_bucket = GetBucketCap();
+    // Write other values important for iteration
+    info.crit_attempts++;
+    info.restore_data(wep);
+}
+
+bool CanWeaponCrit(IClientEntity *wep)
+{
+    // Are we allowed to even use that much?
+    if (g_pLocalPlayer->weapon_mode != weapon_melee && !IsAllowedToWithdrawFromCritBucket(wep, added_per_shot))
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:138]: Well they can, just this function does not matter for them, will add comment

File: src/crits.cpp [hunk 11]
```diff
@@ -1,357 +1,655 @@
...
-        ++number;
+        // Do calculations
+        SimulateNormalShot(wep, added_per_shot);
     }
+    // Restore variables
+    info.restore_data(wep);
+    return shots;
+}
+
+typedef float (*AttribHookFloat_t)(float, const char *, IClientEntity *, void *, bool);
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:162]: i always used typedef so far but sure i can try that

File: src/crits.cpp [hunk 12]
```diff
@@ -1,357 +1,655 @@
...
 }
 
-void unfuck_bucket(IClientEntity *weapon)
+// Server gives us garbage so let's just calc our own
+float GetObservedCritChane()
 {
-    static bool changed;
-    static float last_bucket;
-    static int last_weapon;
+    int normal_damage = cached_damage;
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:183]: Hm true i think i used it in here before and forgot about the variable...

File: src/crits.cpp [hunk 13]
```diff
@@ -1,357 +1,655 @@
...
+static Timer random_crit_refresh{};
+bool random_crits_enabled()
 {
-    int command_number;
-    int init_command;
-    int weapon_entity;
-};
+    static bool return_value;
+    static ConVar *tf_weapon_criticals = g_ICvar->FindVar("tf_weapon_criticals");
+    if (random_crit_refresh.test_and_set(3000))
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:243]: alright

File: src/crits.cpp [hunk 14]
```diff
@@ -1,357 +1,655 @@
...
-bool force_crit(IClientEntity *weapon)
+// We need to store a bunch of data for when we kill someone with a crit
+struct player_status
 {
-    auto command_number = current_user_cmd->command_number;
+    int health{};
+    bool was_jarated{};
+    bool was_markedfordeath{};
+};
+static std::array<player_status, 33> player_status_list{};
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:271]: Yeah we might want to do that, but even then we'll need this sort of array and we'll need to copy a bunch of data into it.

File: src/crits.cpp [hunk 15]
```diff
@@ -1,357 +1,655 @@
...
-    if (crit_experimental && GetWeaponMode() != weapon_melee)
     {
-        cached_calculation.command_number = number;
-        cached_calculation.weapon_entity  = LOCAL_W->m_IDX;
-        if (!crit_legiter)
+        int next_crit = next_crit_tick();
+        if (next_crit != -1)
         {
-            if (number && number != command_number)
+            // We can just force to nearest crit for melee, and sticky launchers apparently
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:294]: WIll do, we can just check "which nearest tick works" and force command_number to it for melee weapons, and for the sticky launcher aswell. They do not seem to work well with other methods

File: src/crits.cpp [hunk 16]
```diff
@@ -1,357 +1,655 @@
...
-                current_user_cmd->buttons |= IN_ATTACK;
         }
     }
-    lastweapon = weapon->GetModel();
-    lastnumber = number;
-    return number != 0;
 }
 
-void handle_melee_crit(IClientEntity *weapon)
+void update_cmds()
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:313]: Sure, will add a comment 

It updates which command numbers are useable to force crits

File: src/crits.cpp [hunk 17]
```diff
@@ -1,357 +1,655 @@
...
 {
-    // Should i crit or save my crits?
-    bool in_range = false;
-    for (int i = 1; i <= g_IEngine->GetMaxClients(); i++)
-    {
-        // We already found out we should crit
-        if (in_range)
-            break;
-        CachedEntity *target = ENTITY(i);
+    // Melee weapons do not matter
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:315]: On melee weapons we do not calculate this

File: src/crits.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (TotallyNotElite) [reply src/crits.cpp:81]: Why not change it to std::clamp tho? It's more clean. Not really our fault source is c++99

File: src/crits.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (TotallyNotElite) [reply src/crits.cpp:315]: I don't want an answer here lol, this should be added to comments instead.

File: src/crits.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (TotallyNotElite) [reply src/crits.cpp:338]: Add to comments

File: src/crits.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (TotallyNotElite) [reply src/crits.cpp:346]: Better comments.

File: src/crits.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (TotallyNotElite) [reply src/crits.cpp:474]: Comments.

File: src/crits.cpp [hunk 18]
```diff
@@ -1,357 +1,655 @@
...
+        crit_bucket          = *(float *) ((uintptr_t) weapon + 0xa38);
+        weapon_seed          = *(unsigned int *) ((uintptr_t) weapon + 0xb3c);
+        unknown1             = *(unsigned int *) ((uintptr_t) weapon + 0xb30);
+        unknown2             = *(unsigned int *) ((uintptr_t) weapon + 0xb34);
+        unknown3             = *(bool *) ((uintptr_t) weapon + 0xb17);
+        unknown4             = *(float *) ((uintptr_t) weapon + 0xb40);
+        crit_attempts        = *(int *) ((uintptr_t) weapon + 0xa3c);
+        crit_count           = *(int *) ((uintptr_t) weapon + 0xa40);
+        observed_crit_chance = *(float *) ((uintptr_t) weapon + 0xbfc);
+        unknown7             = *(bool *) ((uintptr_t) weapon + 0xb18);
```
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:49]: Result:
- Copying whole data causes lag
- Not copying whole data would mean implementing weapon_info somewhere else instead

Solution:
Keep using this method.

File: src/crits.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:123]: The function restores all the weapon data afterwards

File: src/crits.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (TotallyNotElite) [reply src/crits.cpp]: What you changed it to still isn't good. Why is lower "better for us"?

File: src/crits.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (BenCat07) [reply src/crits.cpp:81]: alright

File: src/crits.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (TotallyNotElite) [reply src/crits.cpp:271]: TODO later

File: include/reclasses/CTFPlayerShared.hpp [hunk 19]
```diff
@@ -30,7 +30,7 @@ class CTFPlayerShared
     {
         static auto signature = gSignatures.GetClientSignature("55 89 E5 57 56 53 83 EC 6C 8B 7D ? C7 44 24 ? 0B 00 00 00");
         typedef bool (*IsCritBoosted_t)(CTFPlayerShared *);
-        IsCritBoosted_t IsCritBoosted_fn = (IsCritBoosted_t) signature;
+        static IsCritBoosted_t IsCritBoosted_fn = (IsCritBoosted_t) signature;
```
Comment (TotallyNotElite) [inline include/reclasses/CTFPlayerShared.hpp:33]: Instead of making both static, you could put`signature` on the same line

Comment (BenCat07) [reply include/reclasses/CTFPlayerShared.hpp:33]: Could or should?

**Pull request merged**